### ISSUE TITLE
bottlerocket with mulitple cluster-dns 

### ISSUE THREAD

**Issue by mzupan** [aws/karpenter-provider-aws#4836]
### Description

**Observed Behavior**:

Not able to set mulitple ips for cluster-dns-ip

bottlerocket allows you to set it via

```
# Valid, single IP
[settings.kubernetes]
"cluster-dns-ip" = "10.0.0.1"

# Also valid, multiple nameserver IPs
[settings.kubernetes]
"cluster-dns-ip" = ["10.0.0.1", "10.0.0.2"]
```

Trying to set it like

```
[settings.kubernetes]
"cluster-dns-ip" = ["10.0.0.1", "10.0.0.2"]
```

Gives the following log

```
﻿karpenter-f7cdc49f4-25kbx controller 2023-10-15T02:10:53.850Z	ERROR	controller	Reconciler error	{"commit": "322822a", "controller": "machine.lifecycle", "controllerGroup": "karpenter.sh", "controllerKind": "Machine", "Machine": {"name":"workers-gmsx5"}, "namespace": "", "name": "workers-gmsx5", "reconcileID": "8ace2c6e-45f7-4bb4-84d8-79388dde83b0", "error": "launching machine, creating instance, getting launch template configs, getting launch templates, creating launch template, invalid UserData toml: cannot decode TOML array into struct field bootstrap.BottlerocketKubernetes.ClusterDNSIP of type *string"}
```

Looking at the code it takes a string but even if you do it with a comma it always just has one which is the default kube-dns

https://github.com/aws/karpenter/blob/main/pkg/providers/amifamily/bootstrap/bottlerocket.go#L59

**Expected Behavior**:

be able to have a /etc/resolv.conf in pods like

node-local helps to stop against loading a lot of coredns servers so we aren't rate limited by the vpc resolver. 

```
$ cat /etc/resolv.conf
search testing.svc.cluster.local svc.cluster.local cluster.local ec2.internal
nameserver 169.254.20.10
nameserver 172.20.0.10
```

**Reproduction Steps** (Please include YAML):

Tried both of the following

```
     userData: |
        [settings.kubernetes]
        "cluster-dns-ip" = ["169.254.20.10","172.20.0.10"]
```

and

```
     userData: |
        [settings.kubernetes]
        "cluster-dns-ip" = "169.254.20.10,172.20.0.10"
```

**Versions**:
- Chart Version: 
- Kubernetes Version (`kubectl version`): 

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jonathan-innis** [comment]
This one is a little painful since our support for Bottlerocket merging is driven by our own struct, meaning that if we were to just do something simple like change this `*string` value to `[]string`, we would be breaking users on the newer version. I'd like to be able to just update our type for `ClusterDNS` in the [`bottlerocketsettings.go`](https://github.com/aws/karpenter/blob/main/pkg/providers/amifamily/bootstrap/bottlerocketsettings.go#L50) file to be a `[]string` but I think we're going to have to support both `*string` and `[]string` like Bottlerocket does. 

I haven't dug in enough to understand exactly how to achieve this through a TOML parse, since I doubt that the default toml library in go supports this functionality.

**Comment by jonathan-innis** [comment]
I'm also going to change this to a feature request since we have had support for the ClusterDNSIP feature with Bottlerocket, but it looks like they added the support for this to be a string or an array in [this PR](https://github.com/bottlerocket-os/bottlerocket/pull/2176).

**Comment by njtran** [comment]
Looks like this can be closed out as part of Bottlerocket v1.8.0. https://github.com/bottlerocket-os/bottlerocket/issues/2132#issuecomment-1151807660

Can you confirm @mzupan @z0rc if this works on this version?

**Comment by z0rc** [comment]
@njtran incorrect. Bottlerocket 1.8.0 introduced support of setting cluster-dns as array. Kaprenter still doesn't support this, [previous comment](https://github.com/aws/karpenter-provider-aws/issues/4836#issuecomment-1763271615) still stands.

OP mentions setting cluster-dns-ip via user data, which is another case and isn't supported, see https://github.com/aws/karpenter-provider-aws/issues/5584#issuecomment-1922665533.

**Comment by jseiser** [comment]
Just wanted to add that we would really need this feature.  We deal with Conntrack issues, so node-local DNS would be a easy fix, but without being able to pass 2 IP's we would be creating a new problem where anytime that local cache pod went down, the host would lose DNS access.

**Comment by woehrl01** [comment]
Hi there, just want to follow up here as I just stumbled over the same problem.

I can see that there is already `string[]` used here: https://github.com/aws/karpenter-provider-aws/blob/e0cd8a470ab533c1f6643a763c0c65575a0fd631/pkg/providers/amifamily/bootstrap/bottlerocketsettings.go#L60C3-L60C118

can't we "just" change ClusterDNSIP here https://github.com/aws/karpenter-provider-aws/blob/e0cd8a470ab533c1f6643a763c0c65575a0fd631/pkg/providers/amifamily/bootstrap/bottlerocketsettings.go#L51 to just be of type `string[]` as well and assign the full array here https://github.com/aws/karpenter-provider-aws/blob/e0cd8a470ab533c1f6643a763c0c65575a0fd631/pkg/providers/amifamily/bootstrap/bottlerocket.go#L55

I haven't seen where the merging you have mentioned is happening, maybe you could give me a hint @jonathan-innis 

I'm happy to take a look and create a PR to adress this.

**Comment by jonathan-innis** [comment]
@woehrl01 I think the challenge here is that we have to keep supporting specifying just the string for backwards compatibility. Realistically, I think we could just try merging as a slice and then try merging as a string if that doesn't work. 

In both cases, the backing data structure would still be a slice.

**Comment by radha45** [comment]
Hello Team,

Can you please suggest any workaround for this for now, i am impacted too.

**Comment by demmonico** [comment]
Same here, an old, annoying issue and MR prepared, but not merged already for quite a long time :( 
https://github.com/aws/karpenter-provider-aws/pull/8546

**Comment by oferprtz-gong** [comment]
Hey Team, any news on this? it will help lots and lots of users/companies.
Thanks!
